File: bootpay/src/main/java/smartwork/co/kr/bootpay/PaymentDialog.java
package smartwork.co.kr.bootpay;

import android.app.AlertDialog;
import android.app.DialogFragment;
import android.app.FragmentManager;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.json.JSONObject;

import java.lang.ref.WeakReference;
import java.util.Collection;
import java.util.List;

import smartwork.co.kr.bootpay.model.Item;
import smartwork.co.kr.bootpay.model.Request;

public final class PaymentDialog extends DialogFragment {

    private BootpayWebView bootpay;
    private EventListener listener;
    private Request result;

    /**
     * @see Builder
     * @deprecated
     */
    @Deprecated
    public PaymentDialog() {
        // not allow
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(STYLE_NO_FRAME, android.R.style.Theme_Holo_Light);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.activity_payment, container);
        bootpay = view.findViewById(R.id.bootpay_web);
        afterViewInit();
        setCancelable(false);
        return view;
    }

    private PaymentDialog setOnResponseListener(EventListener listener) {
        this.listener = listener;
        return this;
    }

    private void afterViewInit() {
        if (bootpay != null) {
            bootpay.setData(result)
                    .setDialog(getDialog())
                    .setOnResponseListener(listener);
        }
    }

    private PaymentDialog setData(Request request) {
        this.result = request;
        return this;
    }

    private void tranactionConfirm(String data) {
        bootpay.transactionConfirm(data);
    }

    /**
     * Ironically, it can be opened even have no errors in Any Thread.
     */
    public static final class Builder {
        private WeakReference<FragmentManager> fragmentManager;
        private Request result;
        private EventListener listener;
        private ErrorListener error;
        private DoneListener done;
        private CancelListener cancel;
        private ConfirmListener confirm;
        private AlertDialog networkErrorDialog;
        private PaymentDialog dialog;

        private Builder() {
            // not allow
        }

        public Builder(@NonNull FragmentManager fm) {
            result = new Request();
            fragmentManager = new WeakReference<>(fm);
        }

        public Builder setApplicationId(@NonNull String id) {
            result.setApplication_id(id);
            return this;
        }

        public Builder setPrice(@IntRange(from = 0) int price) {
            result.setPrice(price);
            return this;
        }

        public Builder setPG(@NonNull String pg) {
            result.setPg(pg);
            return this;
        }

        public Builder setName(@NonNull String name) {
            result.setName(name);
            return this;
        }

        public Builder addItem(@NonNull String name, @IntRange(from = 1) int quantity, String primaryKey, @IntRange(from = 0) int price) {
            result.addItem(new Item(name, quantity, primaryKey, price));
            return this;
        }

        public Builder addItem(Item item) {
            result.addItem(item);
            return this;
        }

        public Builder addItems(Collection<Item> items) {
            if (items != null) items.forEach(this::addItem);
            return this;
        }

        public Builder setItems(@NonNull List<Item> items) {
            result.setItems(items);
            return this;
        }

        public Builder setOrderId(@NonNull String orderId) {
            result.setOrder_id(orderId);
            return this;
        }

        public Builder setMethod(@NonNull String method) {
            result.setMethod(method);
            return this;
        }

        Builder setModel(@NonNull Request request) {
            result = request;
            return this;
        }

        Builder setEventListener(@Nullable EventListener listener) {
            this.listener = listener;
            return this;
        }

        public Builder setParams(Object params) {
            result.setParams(params);
            return this;
        }

        Builder setParams(String params) {
            result.setParams(params);
            return this;
        }

        Builder setParams(JSONObject params) {
            result.setParams(params);
            return this;
        }

        public Builder onCancel(@NonNull CancelListener listener) {
            cancel = listener;
            return this;
        }

        public Builder onConfirm(@NonNull ConfirmListener listener) {
            confirm = listener;
            return this;
        }

        public Builder onDone(@NonNull DoneListener listener) {
            done = listener;
            return this;
        }

        public Builder onError(@NonNull ErrorListener listener) {
            error = listener;
            return this;
        }

        /**
         * Must have value:
         *
         * @see Request#application_id { @link https://alf001.bomgil.in/project/app }
         * @see Request#pg { @value "bootpay", "payapp", "danal", "kcp", "inicis" }
         * @see Request#price
         * @see Request#order_id
         */
        public void show() {

            if (isNullOrEmpty(result.getApplication_id()))
                error("Application id is not configured.");

            if (isNullOrEmpty(result.getPg()))
                error("PG is not configured.");

            if (isNullOrEmpty(result.getPrice()))
                error("Price is not configured.");

            if (isNullOrEmpty(result.getOrderId()))
                error("Order id is not configured.");

            if (listener == null && (error == null || cancel == null || confirm == null || done == null))
                error("Must to be required to handle events.");

            dialog = new PaymentDialog()
                    .setData(result)
                    .setOnResponseListener(listener == null ? new EventListener() {
                        @Override
                        public void onError(String message) {
                            if (error != null) error.onError(message);
                        }

                        @Override
                        public void onCancel(String message) {
                            if (cancel != null) cancel.onCancel(message);
                        }

                        @Override
                        public void onConfirmed(String message) {
                            if (confirm != null) confirm.onConfirmed(message);
                        }

                        @Override
                        public void onDone(String message) {
                            if (done != null) done.onDone(message);
                        }
                    } : listener);
            dialog.onCancel(new DialogInterface() {
                @Override
                public void cancel() {
                    dialog.bootpay.destroy();
                    dialog = null;
                    BootpayDialog.finish();

                }

                @Override
                public void dismiss() {
                    dialog.bootpay.destroy();
                    dialog = null;
                    BootpayDialog.finish();
                }
            });
            if (!fragmentManager.get().isDestroyed()) dialog.show(fragmentManager.get(), "dialog");
        }

        public void transactionConfirm(String data) {
            if (dialog != null) dialog.tranactionConfirm(data);
        }

        private boolean error(String message) {
            throw new RuntimeException(message);
        }

        private boolean isNullOrEmpty(String value) {
            return (value == null) || (value.length() <= 0);
        }

        private boolean isNullOrEmpty(int value) {
            return value <= 0;
        }
    }

}

File: bootpay/src/main/java/kr/co/bootpay/PaymentDialog.kt
package kr.co.bootpay

import android.app.DialogFragment
import android.app.FragmentManager
import android.content.DialogInterface
import android.os.Bundle
import android.support.annotation.IntRange
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kr.co.bootpay.PaymentDialog.Builder
import kr.co.bootpay.enums.Method
import kr.co.bootpay.enums.PG
import kr.co.bootpay.model.Item
import kr.co.bootpay.model.Request
import kr.co.bootpay.pref.UserInfo
import org.json.JSONObject

class PaymentDialog
/** @see Builder */
@Deprecated("") constructor() // not allowed
    : DialogFragment() {

    private var result: Request? = null
    private var bootpay: BootpayWebView? = null
    private var listener: EventListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(DialogFragment.STYLE_NO_FRAME, android.R.style.Theme_Holo_Light)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        bootpay = BootpayWebView(inflater.context)
        afterViewInit()
        return bootpay
    }

    private fun setOnResponseListener(listener: EventListener): PaymentDialog {
        this.listener = listener
        return this
    }

    private fun afterViewInit() {
        bootpay?.setData(result)?.setDialog(dialog)?.setOnResponseListener(listener)
    }

    private fun setData(request: Request?): PaymentDialog {
        result = request
        return this
    }

    override fun onDestroyView() {
        UserInfo.update()
        super.onDestroyView()
    }

    private fun transactionConfirm(data: String?) {
        bootpay?.transactionConfirm(data)
    }

    class Builder {
        private var fm: FragmentManager? = null
        private var result: Request = Request()
        private var listener: EventListener? = null
        private var error: ErrorListener? = null
        private var done: DoneListener? = null
        private var cancel: CancelListener? = null
        private var confirm: ConfirmListener? = null
        private var dialog: PaymentDialog? = null

        private constructor() {
            // not allowed
        }

        constructor(fm: FragmentManager?) {
            this.fm = fm
        }

        fun setApplicationId(id: String): Builder {
            result.application_id = id
            return this
        }

        fun setPrice(price: Long): Builder {
            result.price = price
            return this
        }

        fun setPG(pg: String): Builder {
            result.pg = pg
            return this
        }

        fun setPG(pg: PG): Builder {
            result.pg = when (pg) {
                PG.BOOTPAY -> "bootpay"
                PG.PAYAPP  -> "payapp"
                PG.DANAL   -> "danal"
                PG.KCP     -> "kcp"
                PG.INICIS  -> "inicis"
                PG.LGUP  -> "lgup"
                PG.KAKAO  -> "kakao"
                PG.JTNET  -> "jtnet"
                PG.NICEPAY  -> "nicepay"
                PG.PAYCO  -> "payco"
            }
            return this
        }

        fun setName(name: String): Builder {
            result.name = name
            return this
        }

        fun addItem(name: String, @IntRange(from = 1) quantity: Int, primaryKey: String, price: Long): Builder {
            result.addItem(Item(name, quantity, primaryKey, price, "", "", ""))
            return this
        }

        fun addItem(name: String, @IntRange(from = 1) quantity: Int, primaryKey: String, price: Long, cat1: String, cat2: String, cat3: String): Builder {
            result.addItem(Item(name, quantity, primaryKey, price, cat1, cat2, cat3))
            return this
        }

        fun addItem(item: Item): Builder {
            result.addItem(item)
            return this
        }

        fun isShowAgree(show: Boolean): Builder {
            result.isShowAgree = show
            return this
        }

        fun addItems(items: Collection<Item>?): Builder {
            if (items != null) for (i in items) addItem(i)
            return this
        }

        fun setItems(items: MutableList<Item>): Builder {
            result.items = items
            return this
        }

        fun setOrderId(orderId: String): Builder {
            result.order_id = orderId
            return this
        }

        fun setMethod(method: String): Builder {
            result.method = method
            return this
        }

        fun setModel(request: Request): Builder {
            result = request
            return this
        }

        fun setMethod(method: Method): Builder {
            result.method = when (method) {
                Method.CARD        -> "card"
                Method.CARD_SIMPLE -> "card_simple"
                Method.BANK        -> "bank"
                Method.VBANK       -> "vbank"
                Method.PHONE       -> "phone"
                Method.AUTH        -> "auth"
                Method.CARD_REBILL -> "card_rebill"
                Method.EASY        -> "easy"
                Method.SELECT      -> ""
            }
            return this
        }

        fun setEventListener(listener: EventListener?): Builder {
            this.listener = listener
            return this
        }

        fun setParams(params: Any): Builder {
            result.setParams(params)
            return this
        }

        fun setParams(params: String): Builder {
            result.params = params
            return this
        }

        fun setParams(params: JSONObject): Builder {
            result.setParams(params)
            return this
        }

        fun onCancel(listener: CancelListener): Builder {
            cancel = listener
            return this
        }

        fun onConfirm(listener: ConfirmListener): Builder {
            confirm = listener
            return this
        }

        fun onDone(listener: DoneListener): Builder {
            done = listener
            return this
        }

        fun onError(listener: ErrorListener): Builder {
            error = listener
            return this
        }

        fun setUserEmail(email: String): Builder {
            result.userEmail = email
            return this
        }

        fun setUserName(name: String): Builder {
            result.userName = name
            return this
        }

        fun setUserAddr(address: String): Builder {
            result.userAddr = address
            return this
        }

        fun setUserPhone(number: String): Builder {
            result.userPhone = number
            return this
        }

        /**
         * Must have value:
         *
         * @see Request.application_id { @link https://alf001.bomgil.in/project/app }
         *
         * @see Request.pg { @value "bootpay", "payapp", "danal", "kcp", "inicis" }
         *
         * @see Request.price
         *
         * @see Request.orderId
         */
        fun show() {

            if (result.application_id.isEmpty()) error("Application id is not configured.")

            if (result.pg.isEmpty()) error("PG is not configured.")

            if (result.price < 0) error("Price is not configured.")

            if (result.order_id.isEmpty()) error("Order id is not configured.")

            if (listener == null && (error == null || cancel == null || confirm == null || done == null)) error("Must to be required to handle events.")

            dialog = PaymentDialog().setData(result)
                //                    .setTrace(trace)
                //                    .setUserInfo(user)
                .setOnResponseListener(listener ?: object: EventListener {
                    override fun onError(message: String?) {
                        error?.onError(message)
                    }

                    override fun onCancel(message: String?) {
                        cancel?.onCancel(message)
                    }

                    override fun onConfirmed(message: String?) {
                        confirm?.onConfirmed(message)
                    }

                    override fun onDone(message: String?) {
                        done?.onDone(message)
                    }
                })
            dialog?.onCancel(object: DialogInterface {
                override fun cancel() {
                    dialog?.bootpay?.destroy()
                    dialog = null
                    UserInfo.finish()
                    Bootpay.finish()

                }

                override fun dismiss() {
                    dialog?.bootpay?.destroy()
                    dialog = null
                    UserInfo.finish()
                    Bootpay.finish()
                }
            })
            UserInfo.update()
            if (fm?.isDestroyed == false) dialog?.show(fm, "bootpay")
        }

        fun transactionConfirm(data: String?) {
            dialog?.transactionConfirm(data)
        }

        private fun error(message: String?) {
            throw RuntimeException(message)
        }
    }

}
